<!DOCTYPE html>
<html lang="zh-CN">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width">
<meta name="theme-color" content="#222"><meta name="generator" content="Hexo 7.0.0">

  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">



<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.0/css/all.min.css" integrity="sha256-yIDrPSXHZdOZhAqiBP7CKzIwMQmRCJ8UeB8Jo17YC4o=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/animate.css/3.1.1/animate.min.css" integrity="sha256-PR7ttpcvz8qrF57fur/yAx1qXMFJeJFiA6pSzWi0OIE=" crossorigin="anonymous">

<script class="next-config" data-name="main" type="application/json">{"hostname":"fentouxungui.github.io","root":"/","images":"/images","scheme":"Muse","darkmode":false,"version":"8.19.1","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12},"copycode":{"enable":false,"style":null},"fold":{"enable":false,"height":500},"bookmark":{"enable":false,"color":"#222","save":"auto"},"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"stickytabs":false,"motion":{"enable":true,"async":false,"transition":{"menu_item":"fadeInDown","post_block":"fadeIn","post_header":"fadeInDown","post_body":"fadeInDown","coll_header":"fadeInLeft","sidebar":"fadeInUp"}},"i18n":{"placeholder":"搜索...","empty":"没有找到任何搜索结果：${query}","hits_time":"找到 ${hits} 个搜索结果（用时 ${time} 毫秒）","hits":"找到 ${hits} 个搜索结果"}}</script><script src="/js/config.js"></script>

    <meta name="description" content="loom 文件逐渐被广泛用于单细胞数据存储，Seurat Team 研发了 LoomR 包用于处理 loom 文件，并与 Seurat 包无缝接合。本文翻译了 Seurat 网站上对于 LoomR 包的介绍。原文发表于 2018-04-22。  关于 loom 文件单细胞数据量越来越大，计算需求也以指数形式增长。Seurat Team 发现，尽管使用 sparse matrices， 对于分析">
<meta property="og:type" content="article">
<meta property="og:title" content="LoomR">
<meta property="og:url" content="https://fentouxungui.github.io/2019/10/12/LoomR/index.html">
<meta property="og:site_name" content="FenTouXunGui">
<meta property="og:description" content="loom 文件逐渐被广泛用于单细胞数据存储，Seurat Team 研发了 LoomR 包用于处理 loom 文件，并与 Seurat 包无缝接合。本文翻译了 Seurat 网站上对于 LoomR 包的介绍。原文发表于 2018-04-22。  关于 loom 文件单细胞数据量越来越大，计算需求也以指数形式增长。Seurat Team 发现，尽管使用 sparse matrices， 对于分析">
<meta property="og:locale" content="zh_CN">
<meta property="article:published_time" content="2019-10-12T12:43:03.000Z">
<meta property="article:modified_time" content="2020-06-10T07:51:31.000Z">
<meta property="article:author" content="Yongchao Zhang">
<meta property="article:tag" content="R">
<meta property="article:tag" content="seurat">
<meta property="article:tag" content="loomR">
<meta name="twitter:card" content="summary">


<link rel="canonical" href="https://fentouxungui.github.io/2019/10/12/LoomR/">



<script class="next-config" data-name="page" type="application/json">{"sidebar":"","isHome":false,"isPost":true,"lang":"zh-CN","comments":true,"permalink":"https://fentouxungui.github.io/2019/10/12/LoomR/","path":"2019/10/12/LoomR/","title":"LoomR"}</script>

<script class="next-config" data-name="calendar" type="application/json">""</script>
<title>LoomR | FenTouXunGui</title>
  








  <noscript>
    <link rel="stylesheet" href="/css/noscript.css">
  </noscript>
</head>

<body itemscope itemtype="http://schema.org/WebPage" class="use-motion">
  <div class="headband"></div>

  <main class="main">
    <div class="column">
      <header class="header" itemscope itemtype="http://schema.org/WPHeader"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="切换导航栏" role="button">
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <i class="logo-line"></i>
      <p class="site-title">FenTouXunGui</p>
      <i class="logo-line"></i>
    </a>
      <p class="site-subtitle" itemprop="description">个人网站</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger" aria-label="搜索" role="button">
    </div>
  </div>
</div>







</header>
        
  
  <aside class="sidebar">

    <div class="sidebar-inner sidebar-nav-active sidebar-toc-active">
      <ul class="sidebar-nav">
        <li class="sidebar-nav-toc">
          文章目录
        </li>
        <li class="sidebar-nav-overview">
          站点概览
        </li>
      </ul>

      <div class="sidebar-panel-container">
        <!--noindex-->
        <div class="post-toc-wrap sidebar-panel">
            <div class="post-toc animated"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#%E5%85%B3%E4%BA%8E-loom-%E6%96%87%E4%BB%B6"><span class="nav-number">1.</span> <span class="nav-text">关于 loom 文件</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#loomR"><span class="nav-number">2.</span> <span class="nav-text">loomR</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%AE%89%E8%A3%85"><span class="nav-number">2.1.</span> <span class="nav-text">安装</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%85%B3%E4%BA%8E-loom-objects"><span class="nav-number">2.2.</span> <span class="nav-text">关于 loom objects</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E4%B8%8E-loom-objects-%E5%BB%BA%E7%AB%8B%E5%85%B3%E8%81%94"><span class="nav-number">2.3.</span> <span class="nav-text">与 loom objects 建立关联</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#loom-%E6%96%87%E4%BB%B6%E7%BB%93%E6%9E%84"><span class="nav-number">2.4.</span> <span class="nav-text">loom 文件结构</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E4%B8%8E-loom-objects-%E7%9A%84%E4%BA%A4%E4%BA%92"><span class="nav-number">2.5.</span> <span class="nav-text">与 loom objects 的交互</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#loomR-%E4%B8%AD%E7%9A%84-matrices"><span class="nav-number">2.6.</span> <span class="nav-text">loomR 中的 matrices</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#loom-objects-%E6%B7%BB%E5%8A%A0%E6%95%B0%E6%8D%AE"><span class="nav-number">2.7.</span> <span class="nav-text">loom objects 添加数据</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%9F%BA%E4%BA%8E-Chunk-%E7%9A%84%E8%BF%AD%E4%BB%A3"><span class="nav-number">2.8.</span> <span class="nav-text">基于 Chunk 的迭代</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#MARGINs"><span class="nav-number">2.9.</span> <span class="nav-text">MARGINs</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Selective-chunking"><span class="nav-number">2.10.</span> <span class="nav-text">Selective-chunking</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%85%B3%E9%97%AD-loom-object"><span class="nav-number">2.11.</span> <span class="nav-text">关闭 loom object</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#loomR-%E5%92%8C-Seurat"><span class="nav-number">3.</span> <span class="nav-text">loomR 和 Seurat</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#seurat-%E4%B8%8E-loom-%E6%96%87%E4%BB%B6%E7%9A%84%E8%BD%AC%E6%8D%A2"><span class="nav-number">3.1.</span> <span class="nav-text">seurat 与 loom 文件的转换</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Seurat-standard-workflow"><span class="nav-number">3.2.</span> <span class="nav-text">Seurat standard workflow</span></a></li></ol></li></ol></div>
        </div>
        <!--/noindex-->

        <div class="site-overview-wrap sidebar-panel">
          <div class="site-author animated" itemprop="author" itemscope itemtype="http://schema.org/Person">
  <p class="site-author-name" itemprop="name">Yongchao Zhang</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap animated">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
        <a href="/archives/">
          <span class="site-state-item-count">10</span>
          <span class="site-state-item-name">日志</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
          <a href="/categories/">
        <span class="site-state-item-count">3</span>
        <span class="site-state-item-name">分类</span></a>
      </div>
      <div class="site-state-item site-state-tags">
          <a href="/tags/">
        <span class="site-state-item-count">10</span>
        <span class="site-state-item-name">标签</span></a>
      </div>
  </nav>
</div>

        </div>
      </div>
    </div>

    
  </aside>


    </div>

    <div class="main-inner post posts-expand">


  


<div class="post-block">
  
  

  <article itemscope itemtype="http://schema.org/Article" class="post-content" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="https://fentouxungui.github.io/2019/10/12/LoomR/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.gif">
      <meta itemprop="name" content="Yongchao Zhang">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="FenTouXunGui">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="post" itemscope itemtype="http://schema.org/CreativeWork">
      <meta itemprop="name" content="LoomR | FenTouXunGui">
      <meta itemprop="description" content="">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          LoomR
        </h1>

        <div class="post-meta-container">
          <div class="post-meta">
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-calendar"></i>
      </span>
      <span class="post-meta-item-text">发表于</span>

      <time title="创建时间：2019-10-12 20:43:03" itemprop="dateCreated datePublished" datetime="2019-10-12T20:43:03+08:00">2019-10-12</time>
    </span>
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-calendar-check"></i>
      </span>
      <span class="post-meta-item-text">更新于</span>
      <time title="修改时间：2020-06-10 15:51:31" itemprop="dateModified" datetime="2020-06-10T15:51:31+08:00">2020-06-10</time>
    </span>
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-folder"></i>
      </span>
      <span class="post-meta-item-text">分类于</span>
        <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
          <a href="/categories/%E5%8D%95%E7%BB%86%E8%83%9E/" itemprop="url" rel="index"><span itemprop="name">单细胞</span></a>
        </span>
    </span>

  
</div>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody"><blockquote>
<p>loom 文件逐渐被广泛用于单细胞数据存储，Seurat Team 研发了 LoomR 包用于处理 loom 文件，并与 Seurat 包无缝接合。本文翻译了 Seurat 网站上对于 LoomR 包的介绍。<a target="_blank" rel="noopener" href="https://satijalab.org/loomR/loomR_tutorial.html">原文</a>发表于 2018-04-22。</p>
</blockquote>
<h2 id="关于-loom-文件"><a href="#关于-loom-文件" class="headerlink" title="关于 loom 文件"></a>关于 loom 文件</h2><p>单细胞数据量越来越大，计算需求也以指数形式增长。<code>Seurat Team</code> 发现，尽管使用 <code>sparse matrices</code>， 对于分析 &gt; 100,000 数目的细胞，也是一个严峻挑战，主要的困难是在内存中加载存储单细胞数据。<code>HDF5</code>数据格式可以解决这个问题，它不是在内存中存储数据，而是在硬盘上进行有效的数据存储，因而也适用于百万级的细胞数目。</p>
<span id="more"></span>
<p><code>Linnarson lab</code> 开发了<code>loom</code>数据格式，它基于 <code>HDF5</code> 数据格式，可轻松存储单细胞基因组数据和元数据。他们还发布了<code>python API - Lommpy</code>，来处理<code>loom</code>文件。</p>
<p><code>Seurat Team </code>开发了 与<code>Loompy</code>互补的一个 R 包 - <code>LoomR</code>。<code>LoomR</code> 允许在 R 环境中，加载和处理<code>loom</code>文件。本教程会简单介绍 <code>LoomR</code> 的安装、与 <code>LoomR</code> 对象的交互和如何充分利用 <code>LoomR </code>中的内置的<code>chunking mechanisms</code>。最后，介绍如何在 <code>Seurat workflow</code> 中进行初始化，使得 Seurat 可以兼容 <code>loom</code> 文件, 并在将来希望可以让 <code>Seurat</code> 与<code>HDF5</code>完全兼容。</p>
<h2 id="loomR"><a href="#loomR" class="headerlink" title="loomR"></a>loomR</h2><p><code>Github release: https://github.com/mojaveazure/loomR</code></p>
<p><code>CRAN release: forthcoming</code></p>
<h3 id="安装"><a href="#安装" class="headerlink" title="安装"></a>安装</h3><p><strong>前提条件：</strong><br><code>C++ HDF5 library</code> 和  <code>devtools</code></p>
<p><strong>C++ HDF5 library</strong></p>
<table>
<thead>
<tr>
<th>操作系统</th>
<th>命令</th>
</tr>
</thead>
<tbody><tr>
<td>macOS</td>
<td>brew install hdf5</td>
</tr>
<tr>
<td>Debian and Debian-based OSes</td>
<td>sudo apt install libhdf5-dev</td>
</tr>
<tr>
<td>Red Hat-based OSes</td>
<td>sudo dnf install hdf5-devel or sudo yum install hdf5-devel</td>
</tr>
<tr>
<td>Windows</td>
<td>Download precombiled binaries from Mario Annau <a target="_blank" rel="noopener" href="https://note.youdao.com/">here</a></td>
</tr>
</tbody></table>
<p>关于<code>loom</code>的类字段和方法： <code>?loomR::loom</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Install devtools from CRAN</span></span><br><span class="line">install.packages<span class="punctuation">(</span><span class="string">&quot;devtools&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># Use devtools to install hdf5r and loomR from GitHub</span></span><br><span class="line">devtools<span class="operator">::</span>install_github<span class="punctuation">(</span>repo <span class="operator">=</span> <span class="string">&quot;hhoeflin/hdf5r&quot;</span><span class="punctuation">)</span></span><br><span class="line">devtools<span class="operator">::</span>install_github<span class="punctuation">(</span>repo <span class="operator">=</span> <span class="string">&quot;mojaveazure/loomR&quot;</span><span class="punctuation">,</span> ref <span class="operator">=</span> <span class="string">&quot;develop&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Load loomR</span></span><br><span class="line">library<span class="punctuation">(</span>loomR<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<h3 id="关于-loom-objects"><a href="#关于-loom-objects" class="headerlink" title="关于 loom objects"></a>关于 loom objects</h3><p><code>loomR</code> 通过基于<code>R6</code>的 <code>loom</code>类，实现<code>loom API</code>。与 <code>R6 objects</code> 的交互与<code>S4 Object</code>（像 <code>Seurat Object</code>）略微不同。简单来说，<code>R6 fields</code>可使用<code>$</code>访问，而不是<code>@</code>，比如<code>my.object$field.name</code>。<code>R6 methods</code>可被直接调用（并修改<code>Object</code>）,比如<code>my.object$method()</code>。</p>
<h3 id="与-loom-objects-建立关联"><a href="#与-loom-objects-建立关联" class="headerlink" title="与 loom objects 建立关联"></a>与 loom objects 建立关联</h3><p>与标准的 R objects 不同，loom objects 仅仅是与硬盘上的文件建立联系，而不需要把所需要的数据加载到内存中，这样就会有很小的内存占用。可以用<code>connect</code>关联 loom 文件（例子:<a target="_blank" rel="noopener" href="https://www.dropbox.com/s/jpkytoxihm481jb/pbmc.loom?dl=0">pbmc</a>）、用<code>loomR::create</code>创建表达矩阵、或用<code>Convert</code>将已有的<code>Seurat Object</code>转换为 loom 文件。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Connect to the loom file in read/write mode</span></span><br><span class="line">lfile <span class="operator">&lt;-</span> connect<span class="punctuation">(</span>filename <span class="operator">=</span> <span class="string">&quot;pbmc.loom&quot;</span><span class="punctuation">,</span> mode <span class="operator">=</span> <span class="string">&quot;r+&quot;</span><span class="punctuation">)</span></span><br><span class="line">lfile</span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## Class: loom</span></span><br><span class="line"><span class="comment">## Filename: /home/paul/Documents/Satija/pbmc.loom</span></span><br><span class="line"><span class="comment">## Access type: H5F_ACC_RDWR</span></span><br><span class="line"><span class="comment">## Attributes: version, chunks</span></span><br><span class="line"><span class="comment">## Listing:</span></span><br><span class="line"><span class="comment">##        name    obj_type dataset.dims dataset.type_class</span></span><br><span class="line"><span class="comment">##   col_attrs   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span><br><span class="line"><span class="comment">##  col_graphs   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span><br><span class="line"><span class="comment">##      layers   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span><br><span class="line"><span class="comment">##      matrix H5I_DATASET 2700 x 13714          H5T_FLOAT</span></span><br><span class="line"><span class="comment">##   row_attrs   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span><br><span class="line"><span class="comment">##  row_graphs   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span><br></pre></td></tr></table></figure>
<p>loom Object 容器包括 6 个亚 Object：1 个数据（<code>matrix</code>）和 5 个组（<code>layers, row_attrs, col_attrs, row_graphs, and  col_graphs</code>）<a target="_blank" rel="noopener" href="http://linnarssonlab.org/loompy/format/index.html">更多详情</a></p>
<h3 id="loom-文件结构"><a href="#loom-文件结构" class="headerlink" title="loom 文件结构"></a>loom 文件结构</h3><p>loom 文件，无非是一个强加了严格结构的 HDF5 文件。这个结构有助于保持一个无序二进制文件的一致性，并为哪个数据是是什么提供安全性。下面是此文件结构和调用每个数据规则的一个总结。</p>
<p><code>matrix</code> : loom 文件结构的“根”,二维（<code>n</code> 基因和 <code>m</code> 细胞）</p>
<p><code>layers</code> ：<code>matrix</code>数据的可选代表（映射），与<code>matrix</code>有相同的维度</p>
<p><code>row_attrs</code> 和 <code>col_attrs</code> ：分别是行 (基因) 和 列 (细胞)的元数据，肯定是一维或二维的数据，<code>row_attrs</code>第一个维度是 <code>n</code>，<code>col_attrs</code>的第一个维度是<code>m</code>。</p>
<p><code>row_graphs</code> 和 <code>col_graphs</code>： 坐标轴格式的稀疏散点图，每张图都是一组三个等长的数据：a (行索引)、b （列索引）和 w（值）</p>
<h3 id="与-loom-objects-的交互"><a href="#与-loom-objects-的交互" class="headerlink" title="与 loom objects 的交互"></a>与 loom objects 的交互</h3><p>两种方式：<br>使用<code>double subset [[ operator ]]</code> 或 <code>$</code> 符号。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Viewing the `matrix` dataset with the double subset [[ operator You can</span></span><br><span class="line"><span class="comment"># also use the $ sigil, i.e. lfile$matrix</span></span><br><span class="line">lfile<span class="punctuation">[[</span><span class="string">&quot;matrix&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## Class: H5D</span></span><br><span class="line"><span class="comment">## Dataset: /matrix</span></span><br><span class="line"><span class="comment">## Filename: /home/paul/Documents/Satija/pbmc.loom</span></span><br><span class="line"><span class="comment">## Access type: H5F_ACC_RDWR</span></span><br><span class="line"><span class="comment">## Datatype: H5T_IEEE_F64LE</span></span><br><span class="line"><span class="comment">## Space: Type=Simple     Dims=2700 x 13714     Maxdims=Inf x Inf</span></span><br><span class="line"><span class="comment">## Chunk: 32 x 32</span></span><br></pre></td></tr></table></figure>

<p>查看组内数据，使用<code>[[ operator ]]</code>或 <code>$</code> 符号。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Viewing a dataset in the &#x27;col_attrs&#x27; group with the double subset [[</span></span><br><span class="line"><span class="comment"># operator and full UNIX-style path</span></span><br><span class="line">lfile<span class="punctuation">[[</span><span class="string">&quot;col_attrs/cell_names&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## Class: H5D</span></span><br><span class="line"><span class="comment">## Dataset: /col_attrs/cell_names</span></span><br><span class="line"><span class="comment">## Filename: /home/paul/Documents/Satija/pbmc.loom</span></span><br><span class="line"><span class="comment">## Access type: H5F_ACC_RDWR</span></span><br><span class="line"><span class="comment">## Datatype: H5T_STRING &#123;</span></span><br><span class="line"><span class="comment">##       STRSIZE H5T_VARIABLE;</span></span><br><span class="line"><span class="comment">##       STRPAD H5T_STR_NULLTERM;</span></span><br><span class="line"><span class="comment">##       CSET H5T_CSET_ASCII;</span></span><br><span class="line"><span class="comment">##       CTYPE H5T_C_S1;</span></span><br><span class="line"><span class="comment">##    &#125;</span></span><br><span class="line"><span class="comment">## Space: Type=Simple     Dims=2700     Maxdims=Inf</span></span><br><span class="line"><span class="comment">## Chunk: 1024</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Viewing a dataset in the &#x27;row_attrs&#x27; group with S3 $ chaining</span></span><br><span class="line">lfile<span class="operator">$</span>row.attrs<span class="operator">$</span>gene_names</span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## Class: H5D</span></span><br><span class="line"><span class="comment">## Dataset: /row_attrs/gene_names</span></span><br><span class="line"><span class="comment">## Filename: /home/paul/Documents/Satija/pbmc.loom</span></span><br><span class="line"><span class="comment">## Access type: H5F_ACC_RDWR</span></span><br><span class="line"><span class="comment">## Datatype: H5T_STRING &#123;</span></span><br><span class="line"><span class="comment">##       STRSIZE H5T_VARIABLE;</span></span><br><span class="line"><span class="comment">##       STRPAD H5T_STR_NULLTERM;</span></span><br><span class="line"><span class="comment">##       CSET H5T_CSET_ASCII;</span></span><br><span class="line"><span class="comment">##       CTYPE H5T_C_S1;</span></span><br><span class="line"><span class="comment">##    &#125;</span></span><br><span class="line"><span class="comment">## Space: Type=Simple     Dims=13714     Maxdims=Inf</span></span><br><span class="line"><span class="comment">## Chunk: 1024</span></span><br></pre></td></tr></table></figure>
<p>注意，以上命令仅返回存储数据的描述。获取数据本身，须在描述后面添加<code>[ operator ]</code>。</p>
<p>获取数据需要整数索引，或者空的<code>[]</code>来加载全部数据（<code>[,]</code>二维数据），此时，数据才被加载到内存中。这允许仅加载所需数据到内存中。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Access the upper left corner of the data matrix</span></span><br><span class="line">lfile<span class="punctuation">[[</span><span class="string">&quot;matrix&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">5</span><span class="punctuation">,</span> <span class="number">1</span><span class="operator">:</span><span class="number">5</span><span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">##      [,1] [,2] [,3] [,4] [,5]</span></span><br><span class="line"><span class="comment">## [1,]    0    0    0    0    0</span></span><br><span class="line"><span class="comment">## [2,]    0    0    0    0    0</span></span><br><span class="line"><span class="comment">## [3,]    0    0    0    0    0</span></span><br><span class="line"><span class="comment">## [4,]    0    0    0    0    0</span></span><br><span class="line"><span class="comment">## [5,]    0    0    0    0    0</span></span><br></pre></td></tr></table></figure>

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Access the full data matrix (here, using the $ instead of the [[ operator</span></span><br><span class="line"><span class="comment"># to access the matrix)</span></span><br><span class="line">full.matrix <span class="operator">&lt;-</span> lfile<span class="operator">$</span>matrix<span class="punctuation">[</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line"><span class="built_in">dim</span><span class="punctuation">(</span>x <span class="operator">=</span> full.matrix<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## [1]  2700 13714</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Access all gene names</span></span><br><span class="line">gene.names <span class="operator">&lt;-</span> lfile<span class="punctuation">[[</span><span class="string">&quot;row_attrs/gene_names&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">[</span><span class="punctuation">]</span></span><br><span class="line">head<span class="punctuation">(</span>x <span class="operator">=</span> gene.names<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## [1] &quot;AL627309.1&quot;    &quot;AP006222.2&quot;    &quot;RP11-206L10.2&quot; &quot;RP11-206L10.9&quot;</span></span><br><span class="line"><span class="comment">## [5] &quot;LINC00115&quot;     &quot;NOC2L&quot;</span></span><br></pre></td></tr></table></figure>

<p><strong>get.attribute.df</strong></p>
<p>由于元数据存储在loom文件中的多个数据集中，每个 loom Object 会有一个<code>get.attribute.df</code>方法:把多种元数据集组织成一个 data frame，以便使用。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Pull three bits of metadata from the column attributes</span></span><br><span class="line">attrs <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;nUMI&quot;</span><span class="punctuation">,</span> <span class="string">&quot;nGene&quot;</span><span class="punctuation">,</span> <span class="string">&quot;orig.ident&quot;</span><span class="punctuation">)</span></span><br><span class="line">attr.df <span class="operator">&lt;-</span> lfile<span class="operator">$</span>get.attribute.df<span class="punctuation">(</span>MARGIN <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span> attribute.names <span class="operator">=</span> attrs<span class="punctuation">)</span></span><br><span class="line">head<span class="punctuation">(</span>x <span class="operator">=</span> attr.df<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">##                nUMI nGene    orig.ident</span></span><br><span class="line"><span class="comment">## AAACATACAACCAC 2419   779 SeuratProject</span></span><br><span class="line"><span class="comment">## AAACATTGAGCTAC 4903  1352 SeuratProject</span></span><br><span class="line"><span class="comment">## AAACATTGATCAGC 3147  1129 SeuratProject</span></span><br><span class="line"><span class="comment">## AAACCGTGCTTCCG 2639   960 SeuratProject</span></span><br><span class="line"><span class="comment">## AAACCGTGTATGCG  980   521 SeuratProject</span></span><br><span class="line"><span class="comment">## AAACGCACTGGTAC 2163   781 SeuratProject</span></span><br></pre></td></tr></table></figure>
<h3 id="loomR-中的-matrices"><a href="#loomR-中的-matrices" class="headerlink" title="loomR 中的 matrices"></a>loomR 中的 matrices</h3><p><code>Matrix Transposition in loomR</code><br>为了保证最大的I&#x2F;O速度， <code>HDF5 library</code>会转置 matrix 数据。这意味着，列是基因，行是细胞。提升的速度值得去这样做，并且 Seuat team 已经采取了防范措施，会对使用者进行正确的转置。但是，当直接对 data matrices 进行交互时，仍需注意！为了提升loomR 的兼容性，<code>row.attrs</code>仍指基因，<code>col.attrs</code>为细胞。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Print the number of genes</span></span><br><span class="line">lfile<span class="punctuation">[[</span><span class="string">&quot;row_attrs/gene_names&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="operator">$</span>dims</span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## [1] 13714</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Is the number of genes the same as the second dimension (typically</span></span><br><span class="line"><span class="comment"># columns) for the matrix?</span></span><br><span class="line">lfile<span class="punctuation">[[</span><span class="string">&quot;row_attrs/gene_names&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="operator">$</span>dims <span class="operator">==</span> lfile<span class="punctuation">[[</span><span class="string">&quot;matrix&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="operator">$</span>dims<span class="punctuation">[</span><span class="number">2</span><span class="punctuation">]</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## [1] TRUE</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># For the sake of consistency within the single-cell community, we&#x27;ve</span></span><br><span class="line"><span class="comment"># reversed the dimensions for the `shape` field.  As such, the number of</span></span><br><span class="line"><span class="comment"># genes is stored in `lfile$shape[1]`; the number of cells is stored in the</span></span><br><span class="line"><span class="comment"># second field</span></span><br><span class="line">lfile<span class="punctuation">[[</span><span class="string">&quot;row_attrs/gene_names&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="operator">$</span>dims <span class="operator">==</span> lfile<span class="operator">$</span>shape<span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## [1] TRUE`</span></span><br></pre></td></tr></table></figure>
<p>获取部分基因（或细胞）的数据：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Pull gene expression data for all genes, for the first 5 cells Note that</span></span><br><span class="line"><span class="comment"># we&#x27;re using the row position for cells</span></span><br><span class="line">data.subset <span class="operator">&lt;-</span> lfile<span class="punctuation">[[</span><span class="string">&quot;matrix&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">[</span><span class="number">1</span><span class="operator">:</span><span class="number">5</span><span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line"><span class="built_in">dim</span><span class="punctuation">(</span>x <span class="operator">=</span> data.subset<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## [1]     5 13714</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># You can transpose this matrix if you wish to restore the standard</span></span><br><span class="line"><span class="comment"># orientation</span></span><br><span class="line">data.subset <span class="operator">&lt;-</span> t<span class="punctuation">(</span>x <span class="operator">=</span> data.subset<span class="punctuation">)</span></span><br><span class="line"><span class="built_in">dim</span><span class="punctuation">(</span>x <span class="operator">=</span> data.subset<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## [1] 13714     5</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Pull gene expression data for the gene MS4A1 Note that we&#x27;re using the</span></span><br><span class="line"><span class="comment"># column position for genes</span></span><br><span class="line">data.gene <span class="operator">&lt;-</span> lfile<span class="punctuation">[[</span><span class="string">&quot;matrix&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">[</span><span class="punctuation">,</span> lfile<span class="operator">$</span>row.attrs<span class="operator">$</span>gene_names<span class="punctuation">[</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="string">&quot;MS4A1&quot;</span><span class="punctuation">]</span></span><br><span class="line">head<span class="punctuation">(</span>x <span class="operator">=</span> data.gene<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## [1] 0 6 0 0 0 0</span></span><br></pre></td></tr></table></figure>

<h3 id="loom-objects-添加数据"><a href="#loom-objects-添加数据" class="headerlink" title="loom objects 添加数据"></a>loom objects 添加数据</h3><p>添加层数据、基因元数据和细胞元数据到 loom object。对应的方法分别是<code>add.layer</code>, <code>add.row.attribute</code>, 和<code>add.col.attribute</code>，数据为矩阵或向量。比如，添加 ENSEMBL IDs 到基因元数据中：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Generate random ENSEMBL IDs for demonstration purposes</span></span><br><span class="line">ensembl.ids <span class="operator">&lt;-</span> paste0<span class="punctuation">(</span><span class="string">&quot;ENSG0000&quot;</span><span class="punctuation">,</span> <span class="number">1</span><span class="operator">:</span><span class="built_in">length</span><span class="punctuation">(</span>x <span class="operator">=</span> lfile<span class="operator">$</span>row.attrs<span class="operator">$</span>gene_names<span class="punctuation">[</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># Use add.row.attribute to add the IDs Note that if you want to overwrite an</span></span><br><span class="line"><span class="comment"># existing value, set overwrite = TRUE</span></span><br><span class="line">lfile<span class="operator">$</span>add.row.attribute<span class="punctuation">(</span><span class="built_in">list</span><span class="punctuation">(</span>ensembl.id <span class="operator">=</span> ensembl.ids<span class="punctuation">)</span><span class="punctuation">,</span> overwrite <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">lfile<span class="punctuation">[[</span><span class="string">&quot;row_attrs&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## Class: H5Group</span></span><br><span class="line"><span class="comment">## Filename: /home/paul/Documents/Satija/pbmc.loom</span></span><br><span class="line"><span class="comment">## Group: /row_attrs</span></span><br><span class="line"><span class="comment">## Listing:</span></span><br><span class="line"><span class="comment">##        name    obj_type dataset.dims dataset.type_class</span></span><br><span class="line"><span class="comment">##  ensembl.id H5I_DATASET        13714         H5T_STRING</span></span><br><span class="line"><span class="comment">##  gene_names H5I_DATASET        13714         H5T_STRING</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Find the ENSEMBL ID for TCL1A</span></span><br><span class="line">lfile<span class="punctuation">[[</span><span class="string">&quot;row_attrs/ensembl.id&quot;</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">[</span>lfile<span class="operator">$</span>row.attrs<span class="operator">$</span>gene_names<span class="punctuation">[</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="string">&quot;TCL1A&quot;</span><span class="punctuation">]</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## [1] &quot;ENSG00009584&quot;</span></span><br></pre></td></tr></table></figure>
<h3 id="基于-Chunk-的迭代"><a href="#基于-Chunk-的迭代" class="headerlink" title="基于 Chunk 的迭代"></a>基于 Chunk 的迭代</h3><p>为 loom pbject 内置了<code>chunking</code>方法，对于较大的数据，可以以小块形式加载数据。为了优化内存使用，<code>loomR </code>提供 <code>map</code> 和 <code>apply</code> 方法，可当数据分批次读入时进行计算。在后台，使用会使用<code>batch.scan</code>和<code>batch.next</code>方法，进行低水平的小块数据计算。每一种方法都会处理 loom 文件的全部数据，包括<code>row_attrs</code> 和 <code> col_attr</code>中的一维数据。</p>
<p><code>map</code> 方法可以将全部数据分批次加载到内存中，从而实现对所有的基因或细胞进行运算，最终的分析结果会放到内存中。比如计算每个细胞总的 UMI 数目：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Map rowSums to `matrix`, using 500 cells at a time, returning a vector</span></span><br><span class="line">nUMI_map <span class="operator">&lt;-</span> lfile<span class="operator">$</span>map<span class="punctuation">(</span>FUN <span class="operator">=</span> rowSums<span class="punctuation">,</span> MARGIN <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span> chunk.size <span class="operator">=</span> <span class="number">500</span><span class="punctuation">,</span> dataset.use <span class="operator">=</span> <span class="string">&quot;matrix&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    display.progress <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># How long is nUMI_map? It should be equal to the number of cells in</span></span><br><span class="line"><span class="comment"># `matrix`</span></span><br><span class="line"><span class="built_in">length</span><span class="punctuation">(</span>x <span class="operator">=</span> nUMI_map<span class="punctuation">)</span> <span class="operator">==</span> lfile<span class="operator">$</span>matrix<span class="operator">$</span>dims<span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## [1] TRUE</span></span><br></pre></td></tr></table></figure>

<h3 id="MARGINs"><a href="#MARGINs" class="headerlink" title="MARGINs"></a>MARGINs</h3><p>loom object 的许多方法都有<code>MARGIN</code>参数：告诉 loom 文件，对哪个维度进行反复迭代、添加或者抓取数据。为了与 scRNAseq 的 R 分析包保持一致，<code>MARGIN=1</code>表示行或基因，<code>MARGIN=2</code>表示列或细胞。这同样适用于 loom object 的<code>shape</code>字段，<code>index 1 </code>表示基因数目， <code>index 2</code> 表示细胞数目。</p>
<p><code>apply</code> 同<code>map</code>方法类似，除了<code>apply</code>会把结果保存到 loom 文件中，而不是返回给用户。因此，它会使用更少的内存，仅仅占用当前迭代的内存。要把结果保存到硬盘上，需要提供名字，命名必须是完整 UNIX - 风格的路径名。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Apply rowSums to `matrix`, using 500 cells at a time, storing in</span></span><br><span class="line"><span class="comment"># `col_attrs/umi_apply`</span></span><br><span class="line">lfile<span class="operator">$</span>apply<span class="punctuation">(</span>name <span class="operator">=</span> <span class="string">&quot;col_attrs/umi_apply&quot;</span><span class="punctuation">,</span> FUN <span class="operator">=</span> rowSums<span class="punctuation">,</span> MARGIN <span class="operator">=</span> <span class="number">2</span><span class="punctuation">,</span> chunk.size <span class="operator">=</span> <span class="number">500</span><span class="punctuation">,</span> </span><br><span class="line">    dataset.use <span class="operator">=</span> <span class="string">&quot;matrix&quot;</span><span class="punctuation">,</span> display.progress <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> overwrite <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">lfile<span class="operator">$</span>col.attrs<span class="operator">$</span>umi_apply</span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## Class: H5D</span></span><br><span class="line"><span class="comment">## Dataset: /col_attrs/umi_apply</span></span><br><span class="line"><span class="comment">## Filename: /home/paul/Documents/Satija/pbmc.loom</span></span><br><span class="line"><span class="comment">## Access type: H5F_ACC_RDWR</span></span><br><span class="line"><span class="comment">## Datatype: H5T_IEEE_F64LE</span></span><br><span class="line"><span class="comment">## Space: Type=Simple     Dims=2700     Maxdims=Inf</span></span><br><span class="line"><span class="comment">## Chunk: 1024</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Ensure that all values are the same as doing a non-chunked calculation</span></span><br><span class="line"><span class="built_in">all</span><span class="punctuation">(</span>lfile<span class="operator">$</span>col.attrs<span class="operator">$</span>umi_apply<span class="punctuation">[</span><span class="punctuation">]</span> <span class="operator">==</span> rowSums<span class="punctuation">(</span>x <span class="operator">=</span> lfile<span class="operator">$</span>matrix<span class="punctuation">[</span><span class="punctuation">,</span> <span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## [1] TRUE</span></span><br></pre></td></tr></table></figure>

<h3 id="Selective-chunking"><a href="#Selective-chunking" class="headerlink" title="Selective-chunking"></a>Selective-chunking</h3><p><code>map</code> 和<code>apply</code> 方法均支持处理特定值（不是特定数据集中的全部值），比如使用<code>apply</code> 去<code>scale</code>特定基因的表达值（而不是全部基因）。实现方法：传递一个整数向量给<code>index.use</code>参数。对于<code>map</code>，返回一个与<code>index.use</code>等长的向量或矩阵，对于<code>apply</code>，由于所有 loom object 的数据必须要有相等的大小，被跳过的<code>index</code>会返回<code>NAs</code>。</p>
<h3 id="关闭-loom-object"><a href="#关闭-loom-object" class="headerlink" title="关闭 loom object"></a>关闭 loom object</h3><p>由于闭 loom object  是对于文件的一个链接，必须要关掉，才能保证数据被正确的写入到硬盘上。使用<code>close_all</code>方法。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">lfile<span class="operator">$</span>close_all<span class="punctuation">(</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="loomR-和-Seurat"><a href="#loomR-和-Seurat" class="headerlink" title="loomR 和 Seurat"></a>loomR 和 Seurat</h2><p>loom 文件正在逐渐被 seurat 所兼容。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>Seurat<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h3 id="seurat-与-loom-文件的转换"><a href="#seurat-与-loom-文件的转换" class="headerlink" title="seurat 与 loom 文件的转换"></a>seurat 与 loom 文件的转换</h3><p>Seurat 提供转换工具。<code>seurat object to loom file</code>：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Load the pbmc_small dataset included in Seurat</span></span><br><span class="line">data<span class="punctuation">(</span><span class="string">&quot;pbmc_small&quot;</span><span class="punctuation">)</span></span><br><span class="line">pbmc_small</span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## An object of class seurat in project SeuratProject </span></span><br><span class="line"><span class="comment">##  230 genes across 80 samples.</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Convert from Seurat to loom Convert takes and object in &#x27;from&#x27;, a name of</span></span><br><span class="line"><span class="comment"># a class in &#x27;to&#x27;, and, for conversions to loom, a filename</span></span><br><span class="line">pfile <span class="operator">&lt;-</span> Convert<span class="punctuation">(</span>from <span class="operator">=</span> pbmc_small<span class="punctuation">,</span> to <span class="operator">=</span> <span class="string">&quot;loom&quot;</span><span class="punctuation">,</span> filename <span class="operator">=</span> <span class="string">&quot;pbmc_small.loom&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    display.progress <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">pfile</span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## Class: loom</span></span><br><span class="line"><span class="comment">## Filename: /home/paul/Documents/Satija/pbmc_small.loom</span></span><br><span class="line"><span class="comment">## Access type: H5F_ACC_RDWR</span></span><br><span class="line"><span class="comment">## Attributes: version, chunks</span></span><br><span class="line"><span class="comment">## Listing:</span></span><br><span class="line"><span class="comment">##        name    obj_type dataset.dims dataset.type_class</span></span><br><span class="line"><span class="comment">##   col_attrs   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span><br><span class="line"><span class="comment">##  col_graphs   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span><br><span class="line"><span class="comment">##      layers   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span><br><span class="line"><span class="comment">##      matrix H5I_DATASET     80 x 230          H5T_FLOAT</span></span><br><span class="line"><span class="comment">##   row_attrs   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span><br><span class="line"><span class="comment">##  row_graphs   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span><br></pre></td></tr></table></figure>

<h3 id="Seurat-standard-workflow"><a href="#Seurat-standard-workflow" class="headerlink" title="Seurat standard workflow"></a>Seurat standard workflow</h3><p>使用 seurat object 找到 <code>variable genes</code>：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Normalize data, and find variable genes using the typical Seurat workflow</span></span><br><span class="line">pbmc_small <span class="operator">&lt;-</span> NormalizeData<span class="punctuation">(</span>object <span class="operator">=</span> pbmc_small<span class="punctuation">,</span> display.progress <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">pbmc_small <span class="operator">&lt;-</span> FindVariableGenes<span class="punctuation">(</span>object <span class="operator">=</span> pbmc_small<span class="punctuation">,</span> display.progress <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> </span><br><span class="line">    do.plot <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">head<span class="punctuation">(</span>x <span class="operator">=</span> pbmc_small<span class="operator">@</span>hvg.info<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">##         gene.mean gene.dispersion gene.dispersion.scaled</span></span><br><span class="line"><span class="comment">## PCMT1    3.942220        7.751848               2.808417</span></span><br><span class="line"><span class="comment">## PPBP     5.555949        7.652876               1.216898</span></span><br><span class="line"><span class="comment">## LYAR     4.231004        7.577377               1.528749</span></span><br><span class="line"><span class="comment">## VDAC3    4.128322        7.383980               1.296982</span></span><br><span class="line"><span class="comment">## KHDRBS1  3.562833        7.367928               2.476809</span></span><br><span class="line"><span class="comment">## IGLL5    3.758330        7.319567               2.018088</span></span><br></pre></td></tr></table></figure>
<p>然后使用 loom object：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Run the same workflow using the loom object</span></span><br><span class="line">NormalizeData<span class="punctuation">(</span>object <span class="operator">=</span> pfile<span class="punctuation">,</span> overwrite <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> display.progress <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line">FindVariableGenes<span class="punctuation">(</span>object <span class="operator">=</span> pfile<span class="punctuation">,</span> overwrite <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">,</span> display.progress <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># Normalized data goes into the &#x27;norm_data&#x27; layer, variable gene information</span></span><br><span class="line"><span class="comment"># goes into &#x27;row_attrs&#x27; Are the results equal?</span></span><br><span class="line">par<span class="punctuation">(</span>mfrow <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="number">1</span><span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">plot<span class="punctuation">(</span>x <span class="operator">=</span> t<span class="punctuation">(</span>x <span class="operator">=</span> pfile<span class="operator">$</span>layers<span class="operator">$</span>norm_data<span class="punctuation">[</span><span class="punctuation">,</span> <span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">,</span> y <span class="operator">=</span> pbmc_small<span class="operator">@</span>data<span class="punctuation">,</span> main <span class="operator">=</span> <span class="string">&quot;Normalized Data&quot;</span><span class="punctuation">,</span> </span><br><span class="line">    xlab <span class="operator">=</span> <span class="string">&quot;loom&quot;</span><span class="punctuation">,</span> ylab <span class="operator">=</span> <span class="string">&quot;Seurat&quot;</span><span class="punctuation">)</span></span><br><span class="line">plot<span class="punctuation">(</span>x <span class="operator">=</span> pfile<span class="operator">$</span>row.attrs<span class="operator">$</span>gene_means<span class="punctuation">[</span><span class="punctuation">]</span><span class="punctuation">,</span> y <span class="operator">=</span> pbmc_small<span class="operator">@</span>hvg.info<span class="punctuation">[</span>pfile<span class="operator">$</span>row.attrs<span class="operator">$</span>gene_names<span class="punctuation">[</span><span class="punctuation">]</span><span class="punctuation">,</span> </span><br><span class="line">    <span class="string">&quot;gene.mean&quot;</span><span class="punctuation">]</span><span class="punctuation">,</span> main <span class="operator">=</span> <span class="string">&quot;Gene Means&quot;</span><span class="punctuation">,</span> xlab <span class="operator">=</span> <span class="string">&quot;loom&quot;</span><span class="punctuation">,</span> ylab <span class="operator">=</span> <span class="string">&quot;Seurat&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>
<p>最后切记：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">pfile$close_all()</span><br></pre></td></tr></table></figure>


    </div>

    
    
    

    <footer class="post-footer">
          <div class="post-tags">
              <a href="/tags/R/" rel="tag"># R</a>
              <a href="/tags/seurat/" rel="tag"># seurat</a>
              <a href="/tags/loomR/" rel="tag"># loomR</a>
          </div>

        

          <div class="post-nav">
            <div class="post-nav-item">
                <a href="/2019/10/09/linux-basis-4/" rel="prev" title="Shell：文本操作">
                  <i class="fa fa-angle-left"></i> Shell：文本操作
                </a>
            </div>
            <div class="post-nav-item">
                <a href="/2022/07/01/DESeq2%E7%9A%84%E5%BB%BA%E6%A8%A1%E5%8E%9F%E7%90%86%E5%8F%8A%E7%AE%80%E5%8D%95%E7%94%A8%E6%B3%95/" rel="next" title="">
                   <i class="fa fa-angle-right"></i>
                </a>
            </div>
          </div>
    </footer>
  </article>
</div>






</div>
  </main>

  <footer class="footer">
    <div class="footer-inner">

  <div class="copyright">
    &copy; 
    <span itemprop="copyrightYear">2023</span>
    <span class="with-love">
      <i class="fa fa-heart"></i>
    </span>
    <span class="author" itemprop="copyrightHolder">Yongchao Zhang</span>
  </div>
  <div class="powered-by">由 <a href="https://hexo.io/" rel="noopener" target="_blank">Hexo</a> & <a href="https://theme-next.js.org/muse/" rel="noopener" target="_blank">NexT.Muse</a> 强力驱动
  </div>

    </div>
  </footer>

  
  <div class="toggle sidebar-toggle" role="button">
    <span class="toggle-line"></span>
    <span class="toggle-line"></span>
    <span class="toggle-line"></span>
  </div>
  <div class="sidebar-dimmer"></div>
  <div class="back-to-top" role="button" aria-label="返回顶部">
    <i class="fa fa-arrow-up fa-lg"></i>
    <span>0%</span>
  </div>

<noscript>
  <div class="noscript-warning">Theme NexT works best with JavaScript enabled</div>
</noscript>


  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/animejs/3.2.1/anime.min.js" integrity="sha256-XL2inqUJaslATFnHdJOi9GfQ60on8Wx1C2H8DYiN1xY=" crossorigin="anonymous"></script>
<script src="/js/comments.js"></script><script src="/js/utils.js"></script><script src="/js/motion.js"></script><script src="/js/schemes/muse.js"></script><script src="/js/next-boot.js"></script>

  






  





</body>
</html>
